### PR TITLE
NR-74032: Prevent deny-list ajax time slice metrics

### PR DESCRIPTION
**Pull request by patrickhousley**
### Overview

Prevent time slice metrics collection for ajax calls that match an entry within the deny list.

### Related Issue(s)

https://issues.newrelic.com/browse/NR-74032

### Testing

Added test cases for time slice metrics around ajax calls that should match the deny list.

### CODE REVIEW

File: tests/functional/stn/ajax.test.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    resourcePromise = router.expectResources()
+
+    return Promise.all([resourcePromise, clickPromise])
+  })
+  .then(([result]) => {
+    t.equal(result.res.statusCode, 200, 'server responded with 200')
+
+    const body = result.body
+    const harvestBody = JSON.parse(body).res
+    const loadNodes = harvestBody.filter(function (node) { return node.t === 'event' && node.n === 'load' || node.n === 'readystatechange' })
```
Comment (metal-messiah) [inline tests/functional/stn/ajax.test.js:47]: Can you clarify what is being tested here?

Should you also check that nodes aren't created where n = `xmlhttprequest`, `Ajax`, `fetchStart`?

File: tests/functional/stn/ajax.test.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      page_view_timing: {
+        enabled: false
+      }
+    }
+  })
+
+  let loadPromise = browser.safeGet(assetURL)
+  let rumPromise = router.expectRum()
+  let resourcePromise = router.expectResources()
+  const ajaxPromise = router.expectBeaconRequest(
```
Comment (metal-messiah) [inline tests/functional/stn/ajax.test.js:80]: can you use `expectXHRMetrics()` here so that we aren't crossing error payloads? (err, xhr, cm, sm all share that stream, so the above listener prevents those from polluting)

File: tests/functional/xhr/deny-list.test.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    scriptString: `
+      const xhr = new XMLHttpRequest();
+      xhr.open('GET', 'http://example.com/');
+      xhr.send();
+    `
+  }))
+
+  Promise.all([ajaxPromise, loadPromise, rumPromise])
+    .then(([response]) => {
+      const newAjaxPromise = router.expectBeaconRequest(
```
Comment (metal-messiah) [inline tests/functional/xhr/deny-list.test.js:24]: same here, can we use `expectXhrMetrics()` for `xhr` in errors payloads to ensure other payloads don't come through?

File: tests/functional/xhr/deny-list.test.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    },
+    scriptString: `
+      fetch('http://example.com').then(function() {}).catch(function() {})
+    `
+  }))
+
+  Promise.all([ajaxPromise, loadPromise, rumPromise])
+    .then(([response]) => {
+      const newAjaxPromise = router.expectBeaconRequest(
```
Comment (metal-messiah) [inline tests/functional/xhr/deny-list.test.js:59]: here as well

**Comment by metal-messiah** [comment]
Since this PR is unofficially fixing polyfills, would you mind updating the polyfills test to not run the "check failures" action?  it was giving false positives.

File: src/features/metrics/instrument/workers-helper.js [hunk 5]
```diff
@@ -66,6 +66,10 @@ function insertSupportMetrics(report) {
      * @returns Proxy worker that intercepts the original constructor
      */
     function extendWorkerConstructor(origClass, workerType) {
+        if (typeof Proxy === 'undefined') {
```
[the remark this answers is not in the document]
Comment (patrickhousley) [reply src/features/metrics/instrument/workers-helper.js:69]: Correct.

File: tests/functional/stn/ajax.test.js (hunk 2 above)
Comment (patrickhousley) [reply tests/functional/stn/ajax.test.js:80]: I had to use `expectBeaconRequest` so I could set a custom timeout.

File: tests/functional/xhr/deny-list.test.js (hunk 3 above)
Comment (patrickhousley) [reply tests/functional/xhr/deny-list.test.js:24]: I had to use expectBeaconRequest so I could set a custom timeout.

File: tests/functional/stn/ajax.test.js (hunk 1 above)
Comment (patrickhousley) [reply tests/functional/stn/ajax.test.js:47]: This file was renamed. I only changed the file to add the second test below.

File: tests/functional/stn/ajax.test.js (hunk 2 above)
Comment (metal-messiah) [reply tests/functional/stn/ajax.test.js:80]: IMO we should instead just extend all the promises to accept+pass on a custom timeout, because the issue still remains that if the agent happens to send a supportability metric before the xhrMetrics payload, it will be an inconsistent test.

Comment (patrickhousley) [reply tests/functional/stn/ajax.test.js:80]: Agree. That is done in the fastify rewrite of the test server.

**Review by metal-messiah — APPROVED** [review]

File: src/features/ajax/aggregate/index.js [hunk 6]
```diff
@@ -75,13 +75,6 @@ export class Aggregate extends AggregateBase {
 
       handle('bstXhrAgg', ['xhr', hash, params, metrics], undefined, FEATURE_NAMES.sessionTrace, ee)
```
Comment (cwli24) [inline src/features/ajax/aggregate/index.js:76]: Bit of a scope creep, but should the deny-list also block this from getting into session trace too?

Comment (patrickhousley) [reply src/features/ajax/aggregate/index.js:76]: No. We talked about this and we decided that the info being in the session trace was acceptable.

**Review by cwli24 — APPROVED** [review]
